File: Python3/201 - 400/257 Binary Tree Paths.py
# 44ms, 100.0%
class Solution:
    
    def binaryTreePaths(self, root):
        res = []
        if root: self.btp_helper(root, res, [str(root.val)])
        return res
    
    def btp_helper(self, root, res, st):
        if not root.left and not root.right:
            res.append(''.join(st))
        else:
            if root.left: self.btp_helper(root.left, res, st + ['->', str(root.left.val)])
            if root.right: self.btp_helper(root.right, res, st + ['->', str(root.right.val)])
        

File: 201 - 400/257. Binary Tree Paths.cc
// 7ms, 14.77%
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<string> binaryTreePaths(TreeNode *root) {
        vector<string> res;
        if (root) helper(root, res, to_string(root->val));
        return res;
    }

private:
    void helper(TreeNode *root, vector<string> &res, string t) {
        if (!root->left && !root->right) {
            res.push_back(t);
        } else {
            if (root->left) helper(root->left, res, t + "->" + to_string(root->left->val));
            if (root->right) helper(root->right, res, t + "->" + to_string(root->right->val));
        }
    }
};
